File: Java/LinkedList/03_Singly_Circular_LinkedList_Java/SinglyCL_005.java
/*
===============================================================================
Program Name    : Traversing a Singly Circular Linked List
Program Number  : 005
Description     : This program demonstrates the traversal of a Singly Circular
                  Linked List in Java. Three nodes are created dynamically and
                  connected in a circular manner. A temporary reference is used
                  to traverse the list by moving from one node to the next.
                  The traversal shows that after reaching the last node, the
                  control returns to the head node because of the circular link.

Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class and Object
2. Default Constructor
3. Self-Referential Class
4. Dynamic Memory Allocation using new
5. Circular Linked List
6. Head Pointer
7. Traversal using Temporary Reference

Key Learning Points:
- Traversal begins from the head node.
- A temporary reference is used for traversal.
- temp moves to the next node using temp.next.
- The last node points back to the head node.
- Circular linked lists do not contain a NULL link.

===============================================================================
*/
// Program 05 : Traversing Singly Circular Linked List

class node
{
    public int data;
    public node next;

    public node()
    {
        data = 0;
        next = null;
    }
}

class SinglyCL5
{
    public static void main(String A[])
    {
        node head = null;
        node obj1 = null;
        node obj2 = null;

        // Allocate memory
        head = new node();
        obj1 = new node();
        obj2 = new node();

        // Store data
        head.data = 11;
        obj1.data = 21;
        obj2.data = 51;

        // Connect nodes
        head.next = obj1;
        obj1.next = obj2;
        obj2.next = head;

        // Traversal
        node temp = head;

        System.out.println(temp.data);

        temp = temp.next;
        System.out.println(temp.data);

        temp = temp.next;
        System.out.println(temp.data);

        temp = temp.next;
        System.out.println(temp.data);
    }
}
/*
===============================================================================
Expected Output:
===============================================================================

11
21
51
11

===============================================================================
Memory Representation:
===============================================================================

                      head
                       │
                       ▼
              +----+------+     +----+------+     +----+------+
              | 11 |   •------->| 21 |   •------->| 51 |   •----+
              +----+------+     +----+------+     +----+------+   |
                   ^                                              |
                   |______________________________________________|

Traversal:

temp
 │
 ▼
11 → 21 → 51 → 11 → ...

===============================================================================
Explanation:
===============================================================================

The program creates three node objects and connects them to form a Singly
Circular Linked List.

Initially, the temporary reference temp points to the head node.

The traversal is performed by repeatedly executing:

temp = temp.next;

The traversal sequence is:

Step 1 : temp points to node containing 11.
Step 2 : temp moves to node containing 21.
Step 3 : temp moves to node containing 51.
Step 4 : temp moves again and reaches the head node because the last node
         points back to head.

Unlike a Singly Linear Linked List, traversal in a circular linked list does
not terminate at NULL because there is no NULL reference.

===============================================================================
Program Flow:
===============================================================================

Start Program
      │
      ▼
Create Three Nodes
      │
      ▼
Assign Data
      │
      ▼
Connect Nodes Circularly
      │
      ▼
Initialize temp = head
      │
      ▼
Display 11
      │
      ▼
Move temp to Next Node
      │
      ▼
Display 21
      │
      ▼
Move temp to Next Node
      │
      ▼
Display 51
      │
      ▼
Move temp to Next Node
      │
      ▼
Display 11 Again
      │
      ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Defined a self-referential node class.
2. Created three node objects.
3. Assigned data to each node.
4. Connected the nodes in a circular manner.
5. Initialized a temporary reference for traversal.
6. Traversed the linked list using next references.
7. Verified that traversal returns to the head node.

===============================================================================
Time Complexity:
===============================================================================

Traversal Steps : O(1)

Space Complexity: O(1)

===============================================================================
Conclusion:
===============================================================================

Program 005 demonstrates the basic traversal of a Singly Circular Linked
List using a temporary reference variable. Since the last node points back
to the head node, traversal continues cyclically instead of terminating at
NULL. This concept forms the basis for implementing complete traversal and
other operations in a Singly Circular Linked List.

===============================================================================
*/

File: C++/LinkedList/03_Singly_Circular_Linear_LinkedList_CPP/SinglyCL_005.cpp
/*
===============================================================================
Program Name    : Singly Circular Linked List - Manual Traversal
Program Number  : 005
Description     : This program demonstrates how to traverse a Singly Circular
                  Linked List manually using a temporary pointer. It shows how
                  the pointer moves from one node to the next and finally
                  returns back to the first node.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Singly Circular Linked List
2. Self-Referential Structure
3. Dynamic Memory Allocation (new)
4. Pointer Traversal
5. Temporary Pointer

Key Learning Points:
- Traversal starts from the first node.
- A temporary pointer is used for traversal.
- Each node is visited using temp = temp->next.
- In a circular linked list, traversal returns to the first node.

===============================================================================
*/

#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

int main()
{
    PNODE first = NULL;
    PNODE second = NULL;
    PNODE third = NULL;
    PNODE temp = NULL;

    // Allocate memory
    first = new NODE;
    second = new NODE;
    third = new NODE;

    // Store data
    first->data = 11;
    second->data = 21;
    third->data = 51;

    // Connect nodes
    first->next = second;
    second->next = third;
    third->next = first;

    // Traversal
    temp = first;

    cout<<"Traversal of Singly Circular Linked List :\n"<<endl;

    cout<<temp->data<<endl;

    temp = temp->next;
    cout<<temp->data<<endl;

    temp = temp->next;
    cout<<temp->data<<endl;

    temp = temp->next;
    cout<<temp->data<<endl;

    delete first;
    delete second;
    delete third;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Traversal of Singly Circular Linked List :

11
21
51
11

===============================================================================
Memory Representation:
===============================================================================

          first,temp
              │
              ▼
+-------+     +-------+     +-------+
|  11   | --> |  21   | --> |  51   |
+-------+     +-------+     +-------+
     ^                             |
     |_____________________________|

Traversal:

temp
 │
 ▼
11 → 21 → 51 → 11 (Back to First)

===============================================================================
Program Flow:
===============================================================================

Create Three Nodes
        │
        ▼
Connect Nodes Circularly
        │
        ▼
Assign temp = first
        │
        ▼
Display Current Node
        │
        ▼
Move temp = temp->next
        │
        ▼
Reach First Node Again

===============================================================================
Time Complexity:
===============================================================================

Traversal of 3 Nodes : O(3)
General Traversal    : O(n)

Space Complexity     : O(1)

===============================================================================
Conclusion:
===============================================================================

This program demonstrates manual traversal of a Singly Circular Linked List
using a temporary pointer. The traversal moves through each node one by one
and finally returns to the first node because the last node points back to
the beginning of the list.

===============================================================================
*/
